File: swift/HikeTimeCore/Sources/HikeTimeCore/TimeModel.swift
import Foundation

/// Модель времени прохождения. Перенос прототипа один в один:
/// полином Minetti (стоимость метра пути по уклону) + бюджет мощности.
/// Эталонные значения зашиты в тесты и сверены с Python и JS.
public enum TimeModel {

    public static let maxSpeed = 6.0 / 3.6   // м/с — быстрее пешком не ходят
    public static let minSpeed = 0.15        // м/с — ниже уже не ходьба

    /// Стоимость метра пути, Дж/(кг·м), по уклону i = dh/dx.
    /// Minetti et al. 2002; минимум около i = -0.10 — пологий спуск
    /// дешевле ровного. За |i| > 0.45 полином расходится, зажимаем.
    public static func minettiCost(_ slope: Double) -> Double {
        let i = max(-0.45, min(0.45, slope))
        let c = 280.5 * pow(i, 5) - 58.7 * pow(i, 4) - 76.8 * pow(i, 3)
              + 51.9 * pow(i, 2) + 19.6 * i + 2.5
        return max(c, 0.4)
    }
}

/// Сегмент маршрута: длина по горизонтали и перепад, метры.
public struct Segment: Sendable, Codable {
    public let distance: Double
    public let dh: Double
    public init(distance: Double, dh: Double) {
        self.distance = distance
        self.dh = dh
    }
}

/// Энергетическая модель: скорость из посильной мощности.
public struct EnergyModel: Sendable {
    /// Масса тела, кг.
    public var bodyKg: Double
    /// Груз, кг.
    public var loadKg: Double
    /// Ватт на кг массы тела, которые человек держит часами (форма).
    public var powerPerKg: Double
    /// Во сколько раз кг груза дороже кг своего тела.
    public var loadFactor: Double
    /// Множитель стоимости покрытия: тропа 1.0, тундра/кусты 1.55, болото/камни 1.9.
    public var terrain: Double
    /// Нелинейная надбавка за тяжёлый рюкзак (стиль Пандольфа), 0 до калибровки.
    public var loadPenaltyK: Double

    public init(bodyKg: Double = 75, loadKg: Double = 0,
                powerPerKg: Double = 3.6, loadFactor: Double = 1.15,
                terrain: Double = 1.0, loadPenaltyK: Double = 0) {
        self.bodyKg = bodyKg
        self.loadKg = loadKg
        self.powerPerKg = powerPerKg
        self.loadFactor = loadFactor
        self.terrain = terrain
        self.loadPenaltyK = loadPenaltyK
    }

    var effectiveKg: Double { bodyKg + loadKg * loadFactor }
    var powerW: Double { powerPerKg * bodyKg }   // мощность даёт тело

    public func speed(slope: Double) -> Double {
        let rel = loadKg / bodyKg
        let cost = TimeModel.minettiCost(slope) * terrain
                 * (1.0 + loadPenaltyK * rel * rel)
        let v = powerW / (cost * effectiveKg)
        return max(TimeModel.minSpeed, min(TimeModel.maxSpeed, v))
    }

    /// Ходовое время в часах, без привалов.
    public func timeHours(_ segments: [Segment]) -> Double {
        var t = 0.0
        for s in segments where s.distance > 0 {
            t += s.distance / speed(slope: s.dh / s.distance)
        }
        return t / 3600
    }
}

/// Привалы: походный ритм ~50/10 плюс обед на длинных маршрутах.
public struct Breaks: Sendable, Equatable {
    public let totalHours: Double
    public let shortBreaks: Int      // по 10 минут
    public let lunchMinutes: Int

    public init(movingHours: Double) {
        let movMin = movingHours * 60
        let n10 = max(0, Int((movMin / 55).rounded(.up)) - 1)
        let lunch = movMin > 240 ? 30 : 0
        self.shortBreaks = n10
        self.lunchMinutes = lunch
        self.totalHours = movingHours + Double(n10 * 10 + lunch) / 60
    }
}

/// Классические правила — для сравнения и онбординга.
public enum ClassicModels {

    public static func naive(distM: Double, kmh: Double = 4) -> Double {
        distM / 1000 / kmh
    }

    public static func naismithLangmuir(distM: Double, gainM: Double,
                                        lossM: Double, segments: [Segment]) -> Double {
        var h = distM / 1000 / 5 + gainM / 600
        for s in segments where s.dh < 0 && s.distance > 0 {
            let angle = atan(abs(s.dh) / s.distance) * 180 / .pi
            if angle >= 5 && angle <= 12 {
                h -= abs(s.dh) / 300 * (10.0 / 60)
            } else if angle > 12 {
                h += abs(s.dh) / 300 * (10.0 / 60)
            }
        }
        return max(h, 0)
    }

    public static func din33466(distM: Double, gainM: Double, lossM: Double) -> Double {
        let horiz = distM / 1000 / 4
        let vert = gainM / 300 + lossM / 500
        return max(horiz, vert) + min(horiz, vert) / 2
    }

    public static func munter(distM: Double, gainM: Double, lossM: Double,
                              up: Double = 4, flat: Double = 6, down: Double = 10) -> Double {
        let km = distM / 1000
        let total = max(gainM + lossM, 1e-9)
        var t = 0.0
        if gainM > 0 { t += (km * gainM / total + gainM / 100) / up }
        if lossM > 0 { t += (km * lossM / total + lossM / 100) / down }
        if gainM + lossM == 0 { t = km / flat }
        return t
    }

    public static func tobler(_ segments: [Segment]) -> Double {
        var t = 0.0
        for s in segments where s.distance > 0 {
            let slope = s.dh / s.distance
            let kmh = 6 * exp(-3.5 * abs(slope + 0.05))
            t += (s.distance / 1000) / max(kmh, 0.3)
        }
        return t
    }
}

File: model/timemodel.py
"""Модели времени прохождения.

Классические правила (Naismith, DIN, Munter, Tobler) — как опора для сравнения.
Наш кандидат — энергетический: считаем стоимость метра пути по уклону,
задаём посильную мощность, из неё получаем скорость. Вес входит физически,
а не отдельным штрафом сверху.
"""
import math

MAX_SPEED = 6.0 / 3.6      # м/с, быстрее пешком по тропе не ходят
MIN_SPEED = 0.15           # м/с, ниже — уже не ходьба


# ---------- классические правила ----------

def naismith_langmuir(dist_m, gain_m, loss_m, segments=None, flat_kmh=5.0):
    """Naismith: 5 км/ч + 1 час на 600 м набора. Поправки Лангмура на спуск."""
    h = dist_m / 1000.0 / flat_kmh + gain_m / 600.0
    if segments:
        for d, dh in segments:
            if dh >= 0 or d <= 0:
                continue
            angle = math.degrees(math.atan(abs(dh) / d))
            if 5 <= angle <= 12:
                h -= (abs(dh) / 300.0) * (10 / 60.0)   # пологий спуск ускоряет
            elif angle > 12:
                h += (abs(dh) / 300.0) * (10 / 60.0)   # крутой спуск тормозит
    return max(h, 0.0)


def din33466(dist_m, gain_m, loss_m, flat_kmh=4.0, up_mh=300.0, down_mh=500.0):
    """DIN 33466: большее из горизонтального и вертикального плюс половина меньшего."""
    horiz = dist_m / 1000.0 / flat_kmh
    vert = gain_m / up_mh + loss_m / down_mh
    return max(horiz, vert) + min(horiz, vert) / 2.0


def munter(dist_m, gain_m, loss_m, up_rate=4.0, flat_rate=6.0, down_rate=10.0):
    """Метод Мюнтера: единицы = км + метры набора/100, делим на темп."""
    km = dist_m / 1000.0
    t = 0.0
    if gain_m > 0:
        t += (km * gain_m / max(gain_m + loss_m, 1e-9) + gain_m / 100.0) / up_rate
    if loss_m > 0:
        t += (km * loss_m / max(gain_m + loss_m, 1e-9) + loss_m / 100.0) / down_rate
    if gain_m + loss_m == 0:
        t = km / flat_rate
    return t


def tobler(segments):
    """Функция Тоблера: скорость от уклона. Про вес не знает вовсе."""
    t = 0.0
    for d, dh in segments:
        if d <= 0:
            continue
        s = dh / d
        kmh = 6.0 * math.exp(-3.5 * abs(s + 0.05))
        t += (d / 1000.0) / max(kmh, 0.3)
    return t


# ---------- энергетическая модель ----------

def minetti_cost(i):
    """Стоимость метра пути, Дж/(кг·м), по уклону i = dh/dx.

    Minetti et al. 2002. Минимум около i = -0.10: пологий спуск дешевле ровного.
    За пределами +-0.45 полином расходится, поэтому зажимаем.
    """
    i = max(-0.45, min(0.45, i))
    c = (280.5 * i**5 - 58.7 * i**4 - 76.8 * i**3
         + 51.9 * i**2 + 19.6 * i + 2.5)
    return max(c, 0.4)


class EnergyModel:
    """Скорость из посильной мощности.

    power_per_kg  — сколько ватт на кг массы тела человек держит часами (форма)
    load_factor   — во сколько раз килограмм груза дороже килограмма своего тела
    terrain       — множитель стоимости: тропа 1.0, осыпь/болото больше
    """

    def __init__(self, body_kg=75.0, load_kg=0.0,
                 power_per_kg=3.6, load_factor=1.15, terrain=1.0,
                 load_penalty_k=0.0):
        self.body_kg = body_kg
        self.load_kg = load_kg
        self.power_per_kg = power_per_kg
        self.load_factor = load_factor
        self.terrain = terrain
        # Нелинейная надбавка за тяжёлый рюкзак, по образцу члена (L/W)^2 у Пандольфа.
        # По умолчанию 0: подбирать только по реальным трекам, иначе это выдуманное число.
        self.load_penalty_k = load_penalty_k

    @property
    def effective_kg(self):
        return self.body_kg + self.load_kg * self.load_factor

    @property
    def power_w(self):
        # мощность даёт тело, поэтому она от своей массы, а не от общей
        return self.power_per_kg * self.body_kg

    def speed(self, slope):
        rel = self.load_kg / self.body_kg
        cost = minetti_cost(slope) * self.terrain * (1.0 + self.load_penalty_k * rel ** 2)
        v = self.power_w / (cost * self.effective_kg)  # м/с
        return max(MIN_SPEED, min(MAX_SPEED, v))

    def time_hours(self, segments):
        t = 0.0
        for d, dh in segments:
            if d <= 0:
                continue
            t += d / self.speed(dh / d)
        return t / 3600.0


def naive_hours(dist_m, kmh=4.0):
    """То, что показывают обычные приложения: расстояние делить на скорость."""
    return dist_m / 1000.0 / kmh
